foo/lib/config/configloader.py

```diff
@@ -0,0 +1,98 @@
+import os
+from typing import Optional
+
+import boto3
+import yaml
+
+
+class ConfigLoaderException(Exception):
+    pass
+
+
+class Config:
+
+    def __init__(self, config: dict, source: str):
+        self._config = config
+        self._source = source
+
+    @property
+    def source(self):
+        return self._source
+
+    def __repr__(self):
+        return (f'Config(config={self._config!r}'
+                f'       source={self.source!r})')
+
+    def __getattr__(self, name):
+        return self._config[name]
+
+    def __getitem__(self, name):
+        return self._config[name]
+
+    def get(self, name, default=None):
+        return self._config.get(name, default)
+
+
+class ConfigLoader:
+
+    """Loads configuration files.
+
+    Source of the config depends on the currently set environment (either
+    'test' or 'prod', as dictated by the *env* file.
+    """
+
+    S3_BUCKET = 'configs'
+    _instance = None
+    _instance_ready = False
+
+    # Make this class a singleton since usually config is loaded only once.
+    def __new__(cls, *args, **kwargs):
+        if cls._instance is None:
+            cls._instance = super().__new__(cls)
+        return cls._instance
+
+    def __init__(self, defaults: dict = {}):
+        if self._instance_ready:
+            return
+        self.defaults = defaults
+        env: str = open('env').read().strip()
```

Comment:
what if such file does not exist?
I think we should raise appropriate exception and log what the problem is.

Reply:
Actually after a while I realized: do we really need to rely on any files here?
If we need to distinguish between `production` and `test` environments can't we rely on some variable that is specified in tests and in tests only and look for those env-variable to get to know which loader we should use?

Reply:
Or maybe even better solution:
How about extracting both `_local_loader` and `_s3_loader` to a separate functions (not class methods) and by default creating ConfigLoader with `self._handler = sf_loader` set in constructor by default. 
However for test create another class, e.g. `TestConfigLoader` with similar to:
```
class TestConfigLoader(ConfigLoader):
    def __init__(*args, **kwargs):
        loader = kwargs.pop('loader', None)
        super().__init__(*args, **kwargs)
        if loader:
            logger.debug("Replacing handler method in config')
            self._handler = loader
```
I think that such solution would be the best one because production code would know nothing about what is done in tests, and still in tests, we could test exact production loader if needed in the future.